core/server/worker/src/main/java/alluxio/worker/netty/DataServerShortCircuitReadHandler.java

```diff
@@ -0,0 +1,154 @@
+ */
+@NotThreadSafe
+class DataServerShortCircuitReadHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(DataServerShortCircuitReadHandler.class);
+
+  /** The block worker. */
+  private final BlockWorker mBlockWorker;
+  /** The lock Id of the block being read. */
+  private long mLockId;
+
+  /**
+   * Creates an instance of {@link DataServerShortCircuitReadHandler}.
+   *
+   * @param blockWorker the block worker
+   */
+  DataServerShortCircuitReadHandler(BlockWorker blockWorker) {
+    mBlockWorker = blockWorker;
+    mLockId = BlockLockManager.INVALID_LOCK_ID;
+  }
+
+  @Override
+  public void channelRead(ChannelHandlerContext ctx, Object msg) {
+    if (!acceptMessage(msg)) {
+      ctx.fireChannelRead(msg);
+      return;
+    }
+
+    ProtoMessage message = ((RPCProtoMessage) msg).getMessage();
+    if (message.isLocalBlockOpenRequest()) {
+      handleBlockOpenRequest(ctx, message.asLocalBlockOpenRequest());
+    } else {
+      Preconditions.checkState(message.isLocalBlockCloseRequest());
+      handleBlockCloseRequest(ctx, message.asLocalBlockCloseRequest());
+    }
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable throwable) {
+    // The RPC handlers do not throw exceptions. All the exception seen here is either
+    // network exception or some runtime exception (e.g. NullPointerException).
+    LOG.error("Failed to handle RPCs.", throwable);
+    ctx.close();
+  }
+
+  @Override
+  public void channelUnregistered(ChannelHandlerContext ctx) {
+    if (mLockId != BlockLockManager.INVALID_LOCK_ID) {
+      try {
+        mBlockWorker.unlockBlock(mLockId);
+      } catch (BlockDoesNotExistException e) {
+        LOG.warn("Failed to unlock lock {} with error {}.", mLockId, e.getMessage());
+      }
+    }
+  }
+
+  /**
+   * @param msg the message
+   * @return true if the message should be processed in handler
+   */
+  private boolean acceptMessage(Object msg) {
+    return (msg instanceof RPCProtoMessage) && (
+        ((RPCProtoMessage) msg).getMessage().isLocalBlockOpenRequest() || ((RPCProtoMessage) msg)
+            .getMessage().isLocalBlockCloseRequest());
+  }
+
+  /**
+   * Handles {@link Protocol.LocalBlockOpenRequest}. No exceptions should be thrown.
+   *
+   * @param ctx the channel handler context
+   * @param request the local block open request
+   */
+  private void handleBlockOpenRequest(ChannelHandlerContext ctx,
+      Protocol.LocalBlockOpenRequest request) {
+    LOG.debug("Open block: {}.", request);
+    try {
+      // It is a no-op to lock the same block multiple times within the same channel.
+      if (mLockId == BlockLockManager.INVALID_LOCK_ID) {
+        mLockId = mBlockWorker.lockBlock(request.getSessionId(), request.getBlockId());
+      } else {
+        LOG.warn("Lock block {} without releasing previous block lock.", request.getBlockId());
```

Comment:
It won't leak lock but it has the risk that the client will read a block without holding proper lock. I have changed it to return a failure to make it more explicit.